Fix duotone theme cache

ajlende opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Fix #36208 

Duotone filters were getting rendered via a side-effect of the global stylesheet generation. Because the stylesheets were cached, the duotone filter rendering wasn't called when the cache hit.

This adds a separate cache for the SVG filters which have to be rendered in the footer of the document, separate from the stylesheet.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

1. Ensure that your environment doesn't have any of these consts set: https://github.com/WordPress/gutenberg/blob/3e84b5671554281453ea0c635742d6a9e2994351/lib/compat/wordpress-5.9/get-global-styles-and-settings.php#L79-L85
2. Apply a duotone filter to a theme via the theme.json file:

```json
{
    "styles": {
        "blocks": {
            "core/image": {
                "filter": {
                    "duotone": "var(--wp--preset--duotone--default-filter)"
                }
            }
        }
    }
}
```

```json
{
    "settings":  {
        "color":  {
            "duotone":  [
                {
                    "colors": [ "#000", "#B9FB9C" ],
                    "slug": "default-filter",
                    "name": "Default filter"
                }
            ]
        }
    }
}
```
3. See that the duotone filters are loaded on all images and persist on subsequent reloads

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

Bug fix

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ajlende committed:
Refactor out the component parts of rendering a duotone filter

ajlende committed:
Reorganize duotone.php functions

ajlende committed:
Further separate out duotone functions

ajlende committed:
Re-add gutenberg_render_duotone_filter_preset

ajlende committed:
Render SVG filters with separate cache

ajlende committed:
Handle per-block duotone presets

ajlende committed:
Skip intermediate array

ajlende committed:
s/filter_url/filter_property/g

ajlende commented on lib/compat/wordpress-5.9/get-global-styles-and-settings.php:
```diff
@@ -120,3 +128,45 @@ function gutenberg_get_global_stylesheet( $types = array() ) {
+	$supports_theme_json = WP_Theme_JSON_Resolver_Gutenberg::theme_has_support();
+	$supports_link_color = get_theme_support( 'experimental-link-color' );
+
+	// TODO: Which origins are needed for duotone filters?
+	$origins = array( 'core', 'theme', 'user' );
+	if ( ! $supports_theme_json && ! $supports_link_color ) {
+		// In this case we only enqueue the core presets (CSS Custom Properties + the classes).
+		$origins = array( 'core' );
+	} elseif ( ! $supports_theme_json && $supports_link_color ) {
+		// For the legacy link color feature to work, the CSS Custom Properties
+		// should be in scope (either the core or the theme ones).
+		$origins = array( 'core', 'theme' );
+	}
```
@oandregal I ended up adding the SVG filters as a separate set because they can't be loaded with `wp_enqueue_style` like the rest in global-styles.php.

I'm assuming the `$supports_link_color = get_theme_support( 'experimental-link-color' );` isn't needed here, but I wanted to double check with you

oandregal replied:
I've pushed some changes to this.

MaggieCabrera commented:
Thank you for this, so quick! This seems to fix the cache issues I was seeing on Skatepark myself

oandregal commented:
Hey, thanks for creating the PR. This looks a bit involved and I have my brain cycles focused on today's freeze. Would you mind if I come back to this on Monday? It's a bug that we need to fix so we have more leeway than for features.

ajlende commented on lib/block-supports/duotone.php:
```diff
@@ -451,6 +486,8 @@ function gutenberg_render_duotone_support( $block_content, $block ) {
 	wp_add_inline_style( $filter_id, $filter_style );
 	wp_enqueue_style( $filter_id );
 
+	gutenberg_render_duotone_filter( $filter_preset );
```
Previously this was handled by `gutenberg_render_duotone_filter_preset` ([above](https://github.com/WordPress/gutenberg/pull/36236/files#diff-fd626ca219f3d805667d480eabb2d890a5a8d2e015b37f03cf8980edf1ae3d58L433)), but now has to be its own separate call.

ajlende commented on lib/block-supports/duotone.php:
```diff
@@ -290,50 +290,43 @@ function gutenberg_tinycolor_string_to_rgb( $color_str ) {
+ * @param  array $preset Duotone preset value as seen in theme.json.
+ * @return string        Duotone filter CSS id.
  */
-function gutenberg_register_duotone_support( $block_type ) {
```
This whole function was just moved [below](https://github.com/WordPress/gutenberg/pull/36236/files#diff-fd626ca219f3d805667d480eabb2d890a5a8d2e015b37f03cf8980edf1ae3d58R421) without modification.

ajlende commented on lib/block-supports/duotone.php:
```diff
@@ -290,50 +290,43 @@ function gutenberg_tinycolor_string_to_rgb( $color_str ) {
+ * @param  array $preset Duotone preset value as seen in theme.json.
+ * @return string        Duotone SVG filter.
  */
-function gutenberg_render_duotone_filter_preset( $preset ) {
```
Split this into a few functions
- `gutenberg_get_duotone_filter_property` for generating the filter `url()` property
- `gutenberg_get_duotone_filter_svg` for creating SVG
- `gutenberg_render_duotone_filter` for rendering the SVG in the appropriate place

This way the SVG can be generated and saved in `get_svg_filters` [below](https://github.com/WordPress/gutenberg/pull/36236/files#diff-b03597cc3da199e5aa5a94950e8241135904853e7c3b82d758e42744878afae7R1318) which can be cached.

The old `gutenberg_render_duotone_filter_preset` still exists [below](https://github.com/WordPress/gutenberg/pull/36236/files#diff-fd626ca219f3d805667d480eabb2d890a5a8d2e015b37f03cf8980edf1ae3d58R413) for `value_func` and for backwards compatibility.

ajlende commented on lib/block-supports/duotone.php:
```diff
@@ -398,8 +401,40 @@ function () use ( $svg ) {
+ *
+ * @param array $preset Duotone preset value as seen in theme.json.
+ */
+function gutenberg_render_duotone_filter_preset( $preset ) {
```
Kept this function so `value_func` in `PRESETS_METADATA` in [class-wp-theme-json-gutenberg.php](https://github.com/WordPress/gutenberg/pull/36236/files#diff-b03597cc3da199e5aa5a94950e8241135904853e7c3b82d758e42744878afae7) didn't have to change.

As a follow-up it may be able to be removed (deprecated) and replaced with `gutenberg_get_duotone_filter_property` when we can ensure that the SVG filters will be generated elsewhere.

oandregal replied:
Hey, I'm still reviewing the changes to the block supports, so I don't fully understand your comment. However, if we plan/want to stop using `value_func` or the callback it's best to do it now. This is something that wasn't part of 5.8 ([see](https://github.com/WordPress/wordpress-develop/blob/5.8/src/wp-includes/class-wp-theme-json.php#L91)) so we can still change it now without worrying about backward compatibility issues (after 5.9 can be harder).

ajlende reviewed (commented):
Added a few comments to help with reviewing

oandregal commented:
Hey, I wasn't able to look at this yet [for reasons](https://make.wordpress.org/core/2021/11/22/wordpress-5-9-revised-release-schedule/). I hope to be able to look at this when the beta1 is in a better place (probably next week?).

One overall comment I'd like to share is that we need to be careful with changes to code that landed in WordPress. For example, changes to code in `lib/block-supports/duotone.php` that was part of WordPress 5.8 ([see](https://github.com/WordPress/wordpress-develop/blob/5.8/src/wp-includes/block-supports/duotone.php)) needs to either stay the same (in terms of name, parameters, etc) or be deprecated (some examples [in this file](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/class-wp-theme-json-resolver.php#L100)).

ajlende committed:
Fix documentation for gutenberg_render_duotone_filter_preset

ajlende commented:
> Hey, I wasn't able to look at this yet for reasons.

No worries, there is a lot going on for the beta release and this is just a bug fix that can go in any time. Feel free to add me on any reviews for things, especially around the color palette work, if you need some help with reviews.

> One overall comment I'd like to share is that we need to be careful with changes to code that landed in WordPress. For example, changes to code in lib/block-supports/duotone.php that was part of WordPress 5.8 (see) needs to either stay the same (in terms of name, parameters, etc) or be deprecated (some examples in this file).

Yep, all functions that existed before still exist with the same signature. I noticed a documentation problem with one of them that I fixed, and I updated some of my review comments to help point to where they all moved since it isn't obvious from the diff.

oandregal commented:
Hey, I've been looking at this during the morning and I have a couple of minor details that I'll push shortly if you don't mind (where code lives and function names to make sure backporting is easy).

I need to review the duotone block supports a bit more and understand what's new in 5.9 and what existed in 5.8. My current understanding of how duotone works and how this PR changes it:

1. The block styles defined in theme.json that use the filters: `.wp-block-image{ filter: var(--wp--preset--duotone--<slug>)}`. Uses the value directly from theme.json. Work as before.

2. The CSS Custom properties to be used in theme.json: `--wp--preset--duotone--<slug>: url(#wp-duotone-<slug>)`. Uses `gutenberg_render_duotone_filter_preset`. Work as before.

3. The SVGs that use the ID. `<svg id="wp-duotone-<slug>" />`:
	- Before: side-effect of calling `gutenberg_render_duotone_filter_preset` => hooked the SVG generation to `wp_body_open` (right after body tag is opened).
	- After: hooks the SVG generation to either `admin_footer` or `wp_footer` filters (right before the body tag is closed), that uses a new public function `gutenberg_get_global_styles_svg_filters` to render them.

Is this correct? If so, I've got a couple of questions:

- Can we remove the code that render the SVGs using the `wp_body_open` filter if we're using now `admin_footer` or `wp_footer` filters instead?
- I have a vague memory that there were some issues depending on where we rendered the SVGs. Is before the closing body a good place? Does it work for all browsers?

oandregal committed:
Fix formatting issues

oandregal committed:
Use wp_ prefix instead of gutenberg

oandregal committed:
Extract function to own file

oandregal commented:
@ajlende got to test this a bit and this is what I ran into. Env: WordPress 5.8, so I could test this PR.

I've used the TwentyTwentyTwo theme and put this bit under `styles.blocks`:

```json
			"core/image": {
				"filter": {
					"duotone": "var(--wp--preset--duotone--foreground-and-background)"
				}
			}
```

It worked for the front & post editor (reloading the page still showed me the images and they used that filter). The images in the site editor didn't get the filter, though.

Then I've tried with a duotone coming from core:

```json
			"core/image": {
				"filter": {
					"duotone": "var(--wp--preset--duotone--blue-red)"
				}
			}
```

and I ran into the same issue as in `trunk` (the image disappears) in the frontend & post editor. The site editor images loaded but had no filters applied.

oandregal committed:
Fix default presets and clean-up

oandregal commented:
@ajlende I've pushed https://github.com/WordPress/gutenberg/pull/36236/commits/81ae91eac463d281a3fbf9e7453b908e958be50e that fixes the issue with default presets, so the only thing missing would be to make sure that this works in the site editor as well?

oandregal commented on lib/compat/wordpress-5.9/get-global-styles-and-settings.php:
```diff
@@ -156,17 +156,10 @@ function wp_get_global_styles_svg_filters() {
-		// should be in scope (either the core or the theme ones).
-		$origins = array( 'core', 'theme' );
+
+	$origins = array( 'default', 'theme' );
```
The `core` key has changed to `default`, that's why the default presets were not working.

oandregal commented:
I've tested on `trunk` and the filters are not loaded for images in the site editor either. This can be looked independently from this PR.

oandregal committed:
Fix rebase/merge issues

oandregal committed:
Only define function if it does not exist

oandregal committed:
Do not render SVGs as a side effect of the global stylesheet

ajlende committed:
Clarify which duotone filters are rendered in the footer

oandregal committed:
Inline missing function

oandregal commented:
> Can we remove the code that render the SVGs using the wp_body_open filter if we're using now admin_footer or wp_footer filters instead?

Removed at https://github.com/WordPress/gutenberg/pull/36236/commits/2109308fd6bea4b932861bc45dc4ce2c79d12a10 We still need to use `gutenberg_render_duotone_filter` to render the custom filters the user may create by itself and attach to an image.

ajlende committed:
Remove unused $types check

oandregal reviewed (approved):
I've tested this in the following scenarios:

- custom filter added by the user
- default filter used by the theme through theme.json
- theme filter used by the theme through theme.json

And it works well in the post and frontend. There's some issue in the site editor but it exists in `trunk` already, so can be looked up separately.

I've pushed some code changes here, so I'll let @ajlende merge this one when he feels is right, in case some of what I've done has caused havoc somewhere.

ajlende commented:
> I've tested on `trunk` and the filters are not loaded for images in the site editor either. This can be looked independently from this PR.

Yeah, this was a know issue I'm taking another look at now. https://github.com/WordPress/gutenberg/issues/35331

ajlende commented:
We discussed these a bit on Slack, but I'll summarize here for future reference.

> Can we remove the code that render the SVGs using the `wp_body_open` filter if we're using now `admin_footer` or `wp_footer` filters instead?

One of the places we render duotone filters is for the user-defined custom per-block filters and the other is for the filters defined in theme.json. If we can somehow register the custom filters into some global store and access them from `wp_get_global_styles_svg_filters` then we can use the singular footer render. But that seems like it would be a bit complicated, so we'll leave that improvement for another day.

ajlende committed:
Adjust theme filters to render the same place as user-defined filters

ajlende committed:
Inline user-defined render so we have one less global function

ajlende commented:
> I have a vague memory that there were some issues depending on where we rendered the SVGs. Is before the closing body a good place? Does it work for all browsers?

Rendering in `wp_body_open` was a fix for Safari incorrectly rendering filters that appeared in the DOM _after_ the content that they were applied to. Safari won't render filters at all if they're placed in the `head` of the document, so `wp_body_open` is what I landed on.

I updated the footer rendering to be in `wp_body_open` and updated the inline documentation for why that was chosen to make it a little more clear.

scruffian commented:
Thanks for merging this!

Mamaduka commented:
Cherry-picked for 5.9.1.
